### PR TITLE
Support parsing recv package name.

### PR DESCRIPTION
**Pull request by tttoad**
Support parsing recv package name. #46 
Please review the code, I will add test cases if there is no problem.

### CODE REVIEW

File: impl.go [hunk 1]
```diff
@@ -447,7 +459,19 @@ to prevent shell globbing.
...
-	fns, err := funcs(iface, *flagSrcDir, EmitComments(*flagComments))
+	if *flagRecvPkg == "" {
+		//  "   s *Struct   " , receiver: Struct
+		recvs := strings.Split(strings.TrimSpace(recv), " ") // [0]:s [1]:*Struct
+		receiver := strings.TrimPrefix(recvs[len(recvs)-1], "*")
+		pkg, _, err := typeSpec("", receiver, *flagSrcDir)
+		if err == nil {
+			*flagRecvPkg = pkg.Package.Name
+		} else {
+			*flagRecvPkg = "---" // Cannot be matched with any package
```
Comment (josharian) [inline impl.go:470]: Why is `"---"` better than `""`?

File: impl.go [hunk 2]
```diff
@@ -447,7 +459,19 @@ to prevent shell globbing.
 		}
 	}
 
-	fns, err := funcs(iface, *flagSrcDir, EmitComments(*flagComments))
+	if *flagRecvPkg == "" {
+		//  "   s *Struct   " , receiver: Struct
```
Comment (josharian) [inline impl.go:463]: I think that something along these lines simpler and clearer (but keeping your original comments, which I elided):

```go
recvs := strings.Fields(recv)
recv := recvs[len(recvs)-1] // note that this correctly handles "s *Struct" and "*Struct"
recv = strings.TrimPrefix(recv, "*")
```

File: impl.go [hunk 3]
```diff
@@ -447,7 +459,19 @@ to prevent shell globbing.
...
+		recvs := strings.Split(strings.TrimSpace(recv), " ") // [0]:s [1]:*Struct
+		receiver := strings.TrimPrefix(recvs[len(recvs)-1], "*")
+		pkg, _, err := typeSpec("", receiver, *flagSrcDir)
+		if err == nil {
+			*flagRecvPkg = pkg.Package.Name
```
Comment (josharian) [inline impl.go:468]: I'd rather not overwrite the global flag var. Can we use a local var instead?

File: impl.go [hunk 4]
```diff
@@ -284,7 +292,7 @@ var errorInterface = []Func{{
 // funcs returns the set of methods required to implement iface.
 // It is called funcs rather than methods because the
 // function descriptions are functions; there is no receiver.
-func funcs(iface string, srcDir string, comments EmitComments) ([]Func, error) {
+func funcs(iface string, srcDir string, recvPkg string, comments EmitComments) ([]Func, error) {
```
Comment (josharian) [inline impl.go:295]: Probably time to make this

`func funcs(iface, srcDir, recvPkg string, ...)`

File: impl.go [hunk 5]
```diff
@@ -208,6 +212,10 @@ func (p Pkg) fullType(e ast.Expr) string {
 	return p.gofmt(e)
 }
 
+func (p Pkg) inRecvPkg() bool {
```
Comment (josharian) [inline impl.go:215]: Please just inline this, as it is used in only the one place.

**Review by josharian — COMMENTED** [review]
Tests sound lovely, thank you.

File: impl.go (hunk 1 above)
Comment (tttoad) [reply impl.go:470]: I just think maybe pkg.Package.Name is "" (maybe that's not possible). So I don't use the default value of string.

Comment (josharian) [reply impl.go:470]: I don't think that's possible. Let's not worry about that possibility.

File: impl.go (hunk 2 above)
Comment (tttoad) [reply impl.go:463]: Good idea.

File: impl.go (hunk 4 above)
Comment (tttoad) [reply impl.go:295]: Do you mean  `func funcs(iface, srcDir, recvPkg string, comments EmitComments )`?
Or `unc funcs(iface string, srcDir string, recvPkg string, ...interface{} )` ?
Or `func funcs(args FuncsArgs)` (FuncsArgs is struct) ?

File: impl.go (hunk 5 above)
Comment (tttoad) [reply impl.go:215]: I think using inRecvPkg() is more readable and extensible.

File: impl.go (hunk 1 above)
Comment (tttoad) [reply impl.go:472]: OK.

File: impl.go (hunk 3 above)
Comment (tttoad) [reply impl.go]: done.

File: impl.go (hunk 4 above)
Comment (josharian) [reply impl.go]: Sorry, I meant the minimal, aesthetic change:

```go
func funcs(iface, srcDir, recvPkg string, comments EmitComments)
```

File: impl.go [hunk 6]
```diff
@@ -120,6 +121,8 @@ func findInterface(iface string, srcDir string) (path string, id string, err err
 type Pkg struct {
 	*build.Package
 	*token.FileSet
+
+	recvPkg string
```
Comment (josharian) [inline impl.go:125]: A comment here would be helpful

File: impl.go (hunk 5 above)
Comment (josharian) [reply impl.go:215]: Respectfully, I disagree.

File: impl.go [hunk 7]
```diff
@@ -447,7 +459,21 @@ to prevent shell globbing.
...
+	if recvPkg == "" {
+		//  "   s *Struct   " , receiver: Struct
+		recvs := strings.Fields(recv)
+		receiver := recvs[len(recvs)-1] // note that this correctly handles "s *Struct" and "*Struct"
+		receiver = strings.TrimPrefix(recv, "*")
+		pkg, _, err := typeSpec("", receiver, *flagSrcDir)
+		if err == nil {
+			recvPkg = pkg.Package.Name
+		} else {
+			*flagRecvPkg = "" // Cannot be matched with any package
```
Comment (josharian) [inline impl.go:472]: On a second read, I believe that this entire else clause can be deleted. The only way to get here is if `recvPkg` is already the empty string. So we can simply do nothing here.

**Review by josharian — COMMENTED** [review]
With a few minor tweaks, this seems good to go. Unless you planned to add more tests? Thanks for the PR.

File: impl.go (hunk 5 above)
Comment (josharian) [reply impl.go:215]: If you're going to keep this as a separate function, it needs docs.

**Review by josharian — APPROVED** [review]
LGTM, once docs are in.

File: impl.go (hunk 5 above)
Comment (tttoad) [reply impl.go]: > Respectfully, I disagree.
Sorry, didn't notice your reply. Since you disagree, I will delete it.

**Comment by josharian** [comment]
Thanks! I rebased this and fixed the last minute bug intro and pushed as df2fb2f8361a1e1af3f6d102eed2f7b5e0fdc8d1.

**Pull request closed without merging**